src/auth/user-import-builder.ts

```diff
@@ -0,0 +1,436 @@
+    derivedKeyLength?: number;
+  };
+}
+
+
+/** User import record as accepted from developer. */
+export interface UserImportRecord {
+  uid: string;
+  email?: string;
+  emailVerified?: boolean;
+  displayName?: string;
+  phoneNumber?: string;
+  photoURL?: string;
+  disabled?: boolean;
+  metadata?: {
+    lastSignInTime?: string;
+    creationTime?: string;
+  };
+  providerData?: Array<{
+    uid: string,
+    displayName?: string,
+    email?: string,
+    photoURL?: string,
+    providerId: string,
+  }>;
+  customClaims?: object;
+  passwordHash?: Buffer;
+  passwordSalt?: Buffer;
+}
+
+
+/** UploadAccount endpoint request interface. */
+export interface UploadAccountRequest {
+  hashAlgorithm?: string;
+  signerKey?: string;
+  rounds?: number;
+  memoryCost?: number;
+  saltSeparator?: string;
+  cpuMemCost?: number;
+  parallelization?: number;
+  blockSize?: number;
+  dkLen?: number;
+  users?: Array<{
+    localId: string;
+    email?: string;
+    emailVerified?: string;
+    displayName?: string;
+    disabled?: boolean;
+    photoUrl?: string;
+    phoneNumber?: string;
+    providerUserInfo?: Array<{
+      rawId: string;
+      providerId: string;
+      email?: string;
+      displayName?: string;
+      photoUrl?: string;
+    }>;
+    passwordHash?: string;
+    salt?: string;
+    lastLoginAt?: number;
+    createdAt?: number;
+    customAttributes?: string;
+  }>;
+}
+
+
+/** Response object for importUsers operation. */
+export interface UserImportResult {
+  failureCount: number;
+  successCount: number;
+  errors: FirebaseArrayIndexError[];
+}
+
+
+/** Callback function to validate an object. */
+type ValidatorFunction = (data: object) => void;
+
+
+/**
+ * Class that provides a helper for building/validating uploadAccount requests and
+ * UserImportResult responses.
+ */
+export class UserImportBuilder {
+  private requiresHashOptions: boolean;
+  private validatedUsers: object[];
+  private validatedOptions: object;
+  private indexMap: object;
+  private userImportResultErrors: FirebaseArrayIndexError[];
+
+  /**
+   * @param {UserImportRecord[]} users The list of user records to import.
+   * @param {UserImportOptions=} options The import options which includes hashing
+   *     algorithm details.
+   * @param {ValidatorFunction=} userRequestValidator The user request validator function.
+   * @constructor
+   */
+  constructor(
+      private users: UserImportRecord[],
+      private options?: UserImportOptions,
+      private userRequestValidator?: ValidatorFunction) {
+    this.requiresHashOptions = false;
+    this.validatedUsers = [];
+    this.userImportResultErrors = [];
+    this.indexMap = {};
+
+    this.validateAndPopulateUsers();
+    this.validateAndPopulateOptions();
+  }
+
+  /**
+   * Returns the corresponding constructed uploadAccount request.
+   * @return {UploadAccountRequest} The constructed uploadAccount request.
+   */
+  public generateRequest(): UploadAccountRequest {
```

Comment:
I think `buildRequest` is a better name for this.

Reply:
done